File: students/DiegoGom05/homework-4/src/banking_system.js
class BankAccount {
  /**
   * Simula un sistema bancario para pruebas de caja negra.
   * @param {string} accountType - 'Checking', 'Savings', o 'Premium'
   * @param {number} initialBalance - Saldo inicial
   */
  constructor(accountType, initialBalance) {
    this.accountType = accountType;
    this.balance = parseFloat(initialBalance);
    this.state = "Active";
    this.dailyTransferTotal = 0.0;
    this.dailyTransferCount = 0;
  }

  /**
   * Retorna el límite de transferencia diario según el tipo de cuenta.
   */
  getDailyLimit() {
    const limits = {
      Savings: 2000.0,
      Checking: 5000.0,
      Premium: 25000.0,
    };
    return limits[this.accountType] || 0.0;
  }

  /**
   * Realiza una transferencia aplicando validaciones de caja negra.
   */
  transfer(amount) {
    const parsedAmount = parseFloat(amount);

    // Validaciones de Estado
    if (this.state === "Closed") {
      return { success: false, error: "Error: Account closed" };
    }
    if (this.state === "Frozen") {
      return { success: false, error: "Error: Account frozen" };
    }

    // Validaciones del Tipo de Cuenta
    if (!["Savings", "Checking", "Premium"].includes(this.accountType)) {
      return { success: false, error: "Error: Unsupported account type" };
    }

    // Validaciones del Monto
    if (isNaN(parsedAmount) || parsedAmount <= 0) {
      return { success: false, error: "Error: Amount must be positive" };
    }

    // Validación de Fondos Disponibles
    if (parsedAmount > this.balance) {
      return { success: false, error: "Error: Insufficient funds" };
    }

    // Validación de Límite Diario acumulado
    const dailyLimit = this.getDailyLimit();
    if (this.dailyTransferTotal + parsedAmount > dailyLimit) {
      return { success: false, error: "Error: Exceeds daily limit" };
    }

    // Ejecución de la transferencia
    this.balance -= parsedAmount;
    this.dailyTransferTotal += parsedAmount;
    this.dailyTransferCount += 1;

    // Transición de estado automática si el saldo cae por debajo de $100
    if (this.balance < 100.0 && this.state === "Active") {
      this.state = "Suspended";
    }

    return { success: true, error: null };
  }

  /**
   * Deposita dinero y restaura el estado a Active si estaba Suspended.
   */
  deposit(amount) {
    if (this.state === "Closed") {
      return { success: false, error: "Error: Account closed" };
    }

    const parsedAmount = parseFloat(amount);
    if (isNaN(parsedAmount) || parsedAmount <= 0) {
      return { success: false, error: "Error: Amount must be positive" };
    }

    this.balance += parsedAmount;
    if (this.state === "Suspended" && this.balance >= 100.0) {
      this.state = "Active";
    }

    return { success: true, error: null };
  }

  /**
   * Procesa el cobro de mantenimiento mensual según la tabla de decisión.
   */
  processMonthlyFee() {
    let fee = 0.0;

    if (this.accountType === "Savings") {
      if (this.balance <= 100.0) fee = 12.0;
    } else if (this.accountType === "Checking") {
      if (this.balance <= 1000.0) fee = 12.0;
    } else if (this.accountType === "Premium") {
      fee = 0.0;
    }

    if (fee > 0) {
      this.balance -= fee;
      return { success: true, feeCharged: fee };
    }

    return { success: true, feeCharged: 0.0 };
  }

  /**
   * Procesa pagos de servicios con validaciones.
   */
  processBillPayment(payeeId, amount, payeeActive = true) {
    if (!payeeActive || payeeId === "999999") {
      return { success: false, error: "Error: Payee not found" };
    }

    if (payeeId.includes("-") || payeeId.includes("!")) {
      return { success: false, error: "Error: Invalid payee format" };
    }

    return this.transfer(amount);
  }

  /**
   * Cambia manualmente el estado de la cuenta.
   */
  changeState(newState, isApproved = true) {
    if (this.state === "Closed") {
      return { success: false, error: "Error: Account closed" };
    }

    if (newState === "Frozen") {
      this.state = "Frozen";
      return { success: true, state: this.state };
    }

    if (newState === "Active" && this.state === "Frozen") {
      if (isApproved) {
        this.state = "Active";
        return { success: true, state: this.state };
      }
      return { success: false, error: "Error: Unfreeze not approved" };
    }

    if (newState === "Closed") {
      this.state = "Closed";
      return { success: true, state: this.state };
    }

    return { success: false, error: "Error: Invalid transition" };
  }
}

module.exports = BankAccount;

File: students/Luis-045/homework-4/src/banking_system.py
"""SecureBank banking system used for black box testing."""


class BankAccount:
    """Represent a SecureBank account."""

    ACCOUNT_RULES = {
        "Savings": {
            "minimum_balance": 100,
            "monthly_fee": 5,
            "waiver_threshold": 1000,
            "daily_limit": 2000,
        },
        "Checking": {
            "minimum_balance": 0,
            "monthly_fee": 10,
            "waiver_threshold": 5000,
            "daily_limit": 5000,
        },
        "Premium": {
            "minimum_balance": 10000,
            "monthly_fee": 0,
            "waiver_threshold": 0,
            "daily_limit": 50000,
        },
    }

    def __init__(self, account_type, initial_balance):
        if account_type not in self.ACCOUNT_RULES:
            raise ValueError("Invalid account type")

        if initial_balance < 0:
            raise ValueError("Initial balance cannot be negative")

        self.account_type = account_type
        self.balance = float(initial_balance)
        self.state = "Active"
        self.daily_transfer_total = 0.0

    def get_daily_limit(self):
        """Return the daily transfer limit."""
        return self.ACCOUNT_RULES[self.account_type]["daily_limit"]

    def get_minimum_balance(self):
        """Return the minimum required balance."""
        return self.ACCOUNT_RULES[self.account_type]["minimum_balance"]

    def transfer(self, amount):
        """Transfer money from the account."""
        if self.state != "Active":
            return {
                "success": False,
                "error": "Account is not active",
            }

        if amount <= 0:
            return {
                "success": False,
                "error": "Amount must be positive",
            }

        if self.daily_transfer_total + amount > self.get_daily_limit():
            return {
                "success": False,
                "error": "Exceeds daily limit",
            }

        if amount > self.balance:
            return {
                "success": False,
                "error": "Insufficient funds",
            }

        self.balance -= amount
        self.daily_transfer_total += amount

        if self.balance < self.get_minimum_balance():
            self.state = "Suspended"

        return {
            "success": True,
            "balance": self.balance,
        }

    def deposit(self, amount):
        """Deposit money into the account."""
        if self.state == "Closed":
            return {
                "success": False,
                "error": "Account closed",
            }

        if self.state == "Frozen":
            return {
                "success": False,
                "error": "Account frozen",
            }

        if amount <= 0:
            return {
                "success": False,
                "error": "Amount must be positive",
            }

        self.balance += amount

        if (
            self.state == "Suspended"
            and self.balance >= self.get_minimum_balance()
        ):
            self.state = "Active"

        return {
            "success": True,
            "balance": self.balance,
        }

    def freeze(self):
        """Freeze an active account."""
        if self.state == "Closed":
            return False

        self.state = "Frozen"
        return True

    def unfreeze(self):
        """Restore a frozen account."""
        if self.state != "Frozen":
            return False

        self.state = "Active"
        return True

    def close(self):
        """Close the account permanently."""
        self.state = "Closed"
        return True

    def process_monthly_fee(self):
        """Charge the monthly account fee when applicable."""
        rules = self.ACCOUNT_RULES[self.account_type]
        fee = rules["monthly_fee"]
        threshold = rules["waiver_threshold"]

        if self.state == "Closed":
            return {
                "success": False,
                "error": "Account closed",
            }

        if fee == 0 or self.balance > threshold:
            return {
                "success": True,
                "fee_charged": 0,
            }

        self.balance -= fee

        if self.balance < self.get_minimum_balance():
            self.state = "Suspended"

        return {
            "success": True,
            "fee_charged": fee,
        }

    def pay_bill(self, payee, amount):
        """Pay a bill using the account balance."""
        if self.state != "Active":
            return {
                "success": False,
                "error": "Account is not active",
            }

        if not payee or not isinstance(payee, str):
            return {
                "success": False,
                "error": "Invalid payee",
            }

        if amount <= 0:
            return {
                "success": False,
                "error": "Amount must be positive",
            }

        if amount > self.balance:
            return {
                "success": False,
                "error": "Insufficient funds",
            }

        self.balance -= amount

        if self.balance < self.get_minimum_balance():
            self.state = "Suspended"

        return {
            "success": True,
            "balance": self.balance,
        }